x/ccv/provider/keeper/permissionless.go

```diff
@@ -0,0 +1,440 @@
+	}
+	return string(buf), true
+}
+
+// SetClientIdToConsumerId sets the client id associated with this consumer id
+func (k Keeper) SetClientIdToConsumerId(ctx sdk.Context, clientId string, consumerId string) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.ClientIdToConsumerIdKey(clientId), []byte(consumerId))
+}
+
+// DeleteClientIdToConsumerId deletes the client id to consumer id association
+func (k Keeper) DeleteClientIdToConsumerId(ctx sdk.Context, clientId string) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.ClientIdToConsumerIdKey(clientId))
+}
+
+// GetInitializedConsumersReadyToLaunch returns the consumer ids of the pending initialized consumer chains
+// that are ready to launch,  i.e., consumer clients to be created.
+func (k Keeper) GetInitializedConsumersReadyToLaunch(ctx sdk.Context) []string {
+	store := ctx.KVStore(k.storeKey)
+	iterator := storetypes.KVStorePrefixIterator(store, types.ConsumerIdToInitializationRecordKeyPrefix())
+	defer iterator.Close()
+
+	var consumerIds []string
+
+	for ; iterator.Valid(); iterator.Next() {
+		// the `consumerId` resides in the whole key, but we skip the first byte (because it's the `ConsumerIdKey`)
+		// plus 8 more bytes for the `uint64` in the key that contains the length of the `consumerId`
+		consumerId := string(iterator.Key()[1+8:])
+
+		var record types.ConsumerInitializationRecord
+		err := record.Unmarshal(iterator.Value())
+		if err != nil {
+			panic(fmt.Errorf("failed to unmarshal consumer record: %w for consumer id: %s", err, consumerId))
+		}
+
+		// if current block time is equal to or after spawnTime, and the chain is initialized, we can launch the chain
+		phase, found := k.GetConsumerIdToPhase(ctx, consumerId)
+		if !ctx.BlockTime().Before(record.SpawnTime) && (found && phase == Initialized) {
+			consumerIds = append(consumerIds, consumerId)
+		}
+	}
+
+	return consumerIds
+}
+
+func (k Keeper) LaunchConsumer(ctx sdk.Context, consumerId string) error {
+	err := k.CreateConsumerClient(ctx, consumerId)
+	if err != nil {
+		return err
+	}
+
+	consumerGenesis, found := k.GetConsumerGenesis(ctx, consumerId)
+	if !found {
+		return errorsmod.Wrapf(types.ErrNoConsumerGenesis, "consumer genesis could not be found")
+	}
+
+	if len(consumerGenesis.Provider.InitialValSet) == 0 {
+		return errorsmod.Wrapf(types.ErrInvalidConsumerGenesis, "consumer genesis initial validator set is empty - no validators opted in")
+	}
+
+	// The cached context is created with a new EventManager so we merge the event
+	// into the original context
+	ctx.EventManager().EmitEvents(ctx.EventManager().Events())
+	return nil
+}
+
+func (k Keeper) UpdateConsumer(ctx sdk.Context, consumerId string) error {
+	phase, found := k.GetConsumerIdToPhase(ctx, consumerId)
+	if !found || phase == Stopped {
+		return errorsmod.Wrapf(types.ErrInvalidPhase,
+			"cannot update stopped or not existing chain: %s", consumerId)
+	}
+
+	updateRecord, found := k.GetConsumerIdToUpdateRecord(ctx, consumerId)
```

Comment:
Deprecated  the top N, validators power cap, validator set cap, etc. state in favor of the `ConsumerUpdateRecord`. However, I still have not deprecated the allowlist, and denylist state. It makes sense in my view to keep those stored in parallel with `ConsumerUpdateRecord` because we can easily access them as `ProviderConsAddr`esses and not merely as `string`s. Additionally, it's easier to query whether a validator is allowlisted without having to traverse through a list of allowlisted addresses every time.

Reply:
Does that mean that every time we need to add a new power shaping config we'll need to migrate the existing `ConsumerUpdateRecords` ?